core/src/server.rs

```diff
@@ -133,172 +91,321 @@ impl SnapshotUrgency {
-        Some(Snapshot { versions_since, .. }) => {
-            SnapshotUrgency::for_versions_since(config, versions_since)
+    /// Implementation of the GetChildVersion protocol transaction.
+    pub fn get_child_version(
+        &self,
+        client_id: ClientId,
+        parent_version_id: VersionId,
+    ) -> Result<GetVersionResult, ServerError> {
+        let mut txn = self.storage.txn()?;
+        let client = txn
+            .get_client(client_id)?
+            .ok_or(ServerError::NoSuchClient)?;
+
+        // If a version with parentVersionId equal to the requested parentVersionId exists, it is
+        // returned.
+        if let Some(version) = txn.get_version_by_parent(client_id, parent_version_id)? {
+            return Ok(GetVersionResult::Success {
+                version_id: version.version_id,
+                parent_version_id: version.parent_version_id,
+                history_segment: version.history_segment,
+            });
         }
-    };
 
-    Ok((
-        AddVersionResult::Ok(version_id),
-        std::cmp::max(time_urgency, version_urgency),
-    ))
-}
+        // Return NotFound if an AddVersion with this parent_version_id would succeed, and
+        // otherwise return Gone.
+        //
+        // AddVersion will succeed if either
+        //  - the requested parent version is the latest version; or
+        //  - there is no latest version, meaning there are no versions stored for this client
+        Ok(
+            if client.latest_version_id == parent_version_id
+                || client.latest_version_id == NIL_VERSION_ID
+            {
+                GetVersionResult::NotFound
+            } else {
+                GetVersionResult::Gone
+            },
+        )
+    }
 
-/// Implementation of the AddSnapshot protocol transaction
-pub fn add_snapshot<'a>(
-    mut txn: Box<dyn StorageTxn + 'a>,
-    _config: &ServerConfig,
-    client_id: ClientId,
-    client: Client,
-    version_id: VersionId,
-    data: Vec<u8>,
-) -> anyhow::Result<()> {
-    log::debug!(
-        "add_snapshot(client_id: {}, version_id: {})",
-        client_id,
-        version_id,
-    );
-
-    // NOTE: if the snapshot is rejected, this function logs about it and returns
-    // Ok(()), as there's no reason to report an errot to the client / user.
-
-    let last_snapshot = client.snapshot.map(|snap| snap.version_id);
-    if Some(version_id) == last_snapshot {
+    /// Implementation of the AddVersion protocol transaction
+    pub fn add_version(
+        &self,
+        client_id: ClientId,
+        parent_version_id: VersionId,
+        history_segment: HistorySegment,
+    ) -> Result<(AddVersionResult, SnapshotUrgency), ServerError> {
+        let mut txn = self.storage.txn()?;
+        let client = txn
+            .get_client(client_id)?
+            .ok_or(ServerError::NoSuchClient)?;
+
+        log::debug!(
+            "add_version(client_id: {}, parent_version_id: {})",
```

Comment:
Depending on your taste, you might prefer `log::debug!("add_version(client_id: {client_id}, parent_version_id: {parent_version_id})");` instead.